File: src/main/java/org/example/kalkulationsprogramm/repository/BelegKostenstellenAnteilRepository.java
package org.example.kalkulationsprogramm.repository;

import org.example.kalkulationsprogramm.domain.BelegKostenstellenAnteil;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.util.List;

public interface BelegKostenstellenAnteilRepository extends JpaRepository<BelegKostenstellenAnteil, Long> {

    @Query("SELECT a FROM BelegKostenstellenAnteil a WHERE a.beleg.id = :belegId ORDER BY a.id ASC")
    List<BelegKostenstellenAnteil> findByBelegId(@Param("belegId") Long belegId);

    @Query("SELECT a FROM BelegKostenstellenAnteil a "
            + "JOIN FETCH a.beleg b "
            + "JOIN FETCH a.kostenstelle ks "
            + "LEFT JOIN FETCH a.zugeordnetVon "
            + "WHERE ks.id = :kostenstelleId "
            + "AND b.status <> org.example.kalkulationsprogramm.domain.BelegStatus.VERWORFEN "
            + "AND NOT EXISTS (SELECT d.id FROM LieferantDokument d WHERE d.beleg = b) "
            + "ORDER BY b.belegDatum DESC, b.uploadDatum DESC, a.id DESC")
    List<BelegKostenstellenAnteil> findByKostenstelleIdEager(@Param("kostenstelleId") Long kostenstelleId);

    @Modifying
    @Query("DELETE FROM BelegKostenstellenAnteil a WHERE a.beleg.id = :belegId")
    void deleteByBelegId(@Param("belegId") Long belegId);

    /**
     * Alle Anteile auf einer bestimmten Kostenstelle, deren Streckungs-Zeitraum
     * das gegebene Jahr enthaelt. Wird vom Verrechnungslohn-Rechner verwendet,
     * um auch gestreckte Bar-Belege in die Gemeinkosten einzubeziehen.
     */
    @Query("SELECT a FROM BelegKostenstellenAnteil a "
            + "WHERE a.kostenstelle.id = :kostenstelleId "
            + "AND (a.streckungStartJahr IS NULL OR a.streckungStartJahr <= :jahr) "
            + "AND (a.streckungStartJahr IS NULL OR a.streckungJahre IS NULL "
            + "     OR a.streckungJahre <= 1 OR :jahr < a.streckungStartJahr + a.streckungJahre)")
    List<BelegKostenstellenAnteil> findAktiveByKostenstelleUndJahr(
            @Param("kostenstelleId") Long kostenstelleId,
            @Param("jahr") int jahr);

    /**
     * Alle Anteile, die im gegebenen Jahr in den Gemeinkostentopf einfliessen:
     * Beleg muss VALIDIERT sein, Kostenstelle muss als Fixkosten markiert sein,
     * und der Streckungs-Zeitraum muss das Jahr abdecken. Wird vom
     * Verrechnungslohn-Rechner verwendet — ersetzt findAll() + Java-Filter,
     * das mit wachsender Beleg-Tabelle nicht skaliert.
     *
     * <p>Semantik exakt wie {@link BelegKostenstellenAnteil#isStreckungAktivFuerJahr(int)}:</p>
     * <ul>
     *   <li>Nicht gestreckt (streckungJahre &le; 1): matched <strong>nur</strong> wenn
     *       {@code streckungStartJahr == :jahr} oder {@code streckungStartJahr IS NULL}
     *       (NULL = "kein konkretes Jahr gesetzt" = unbeschraenkt, Altdaten-Fall).</li>
     *   <li>Gestreckt (streckungJahre &gt; 1): matched fuer Jahre im Intervall
     *       {@code [startJahr, startJahr+jahre)}.</li>
     * </ul>
     * <p>Wichtig: ohne den Year-Match wuerde ein 2024er Bar-Beleg auch in 2025
     * und 2026 wieder in den Gemeinkostentopf einfliessen — Verrechnungslohn-
     * Verfaelschung.</p>
     */
    @Query("SELECT a FROM BelegKostenstellenAnteil a "
            + "JOIN FETCH a.kostenstelle ks "
            + "JOIN FETCH a.beleg b "
            + "WHERE b.status = org.example.kalkulationsprogramm.domain.BelegStatus.VALIDIERT "
            + "AND ks.istFixkosten = true "
            + "AND ("
            + "  ((a.streckungJahre IS NULL OR a.streckungJahre <= 1) "
            + "   AND (a.streckungStartJahr IS NULL OR a.streckungStartJahr = :jahr))"
            + "  OR "
            + "  (a.streckungJahre > 1 AND a.streckungStartJahr IS NOT NULL "
            + "   AND :jahr >= a.streckungStartJahr "
            + "   AND :jahr < a.streckungStartJahr + a.streckungJahre)"
            + ")")
    List<BelegKostenstellenAnteil> findAktiveFixkostenAnteileImJahr(@Param("jahr") int jahr);
}

File: src/main/kotlin/org/example/kalkulationsprogramm/repository/BelegKostenstellenAnteilRepository.kt
package org.example.kalkulationsprogramm.repository

import org.example.kalkulationsprogramm.domain.BelegKostenstellenAnteil
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Modifying
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param

interface BelegKostenstellenAnteilRepository : JpaRepository<BelegKostenstellenAnteil, Long> {
    @Query("SELECT a FROM BelegKostenstellenAnteil a WHERE a.beleg.id = :belegId ORDER BY a.id ASC")
    fun findByBelegId(@Param("belegId") belegId: Long?): List<BelegKostenstellenAnteil>

    @Query(
        "SELECT a FROM BelegKostenstellenAnteil a " +
            "JOIN FETCH a.beleg b " +
            "JOIN FETCH a.kostenstelle ks " +
            "LEFT JOIN FETCH a.zugeordnetVon " +
            "WHERE ks.id = :kostenstelleId " +
            "AND b.status <> org.example.kalkulationsprogramm.domain.BelegStatus.VERWORFEN " +
            "AND NOT EXISTS (SELECT d.id FROM LieferantDokument d WHERE d.beleg = b) " +
            "ORDER BY b.belegDatum DESC, b.uploadDatum DESC, a.id DESC",
    )
    fun findByKostenstelleIdEager(
        @Param("kostenstelleId") kostenstelleId: Long?,
    ): List<BelegKostenstellenAnteil>

    @Modifying
    @Query("DELETE FROM BelegKostenstellenAnteil a WHERE a.beleg.id = :belegId")
    fun deleteByBelegId(@Param("belegId") belegId: Long?)

    @Query(
        "SELECT a FROM BelegKostenstellenAnteil a " +
            "WHERE a.kostenstelle.id = :kostenstelleId " +
            "AND (a.streckungStartJahr IS NULL OR a.streckungStartJahr <= :jahr) " +
            "AND (a.streckungStartJahr IS NULL OR a.streckungJahre IS NULL " +
            "     OR a.streckungJahre <= 1 OR :jahr < a.streckungStartJahr + a.streckungJahre)",
    )
    fun findAktiveByKostenstelleUndJahr(
        @Param("kostenstelleId") kostenstelleId: Long?,
        @Param("jahr") jahr: Int,
    ): List<BelegKostenstellenAnteil>

    @Query(
        "SELECT a FROM BelegKostenstellenAnteil a " +
            "JOIN FETCH a.kostenstelle ks " +
            "JOIN FETCH a.beleg b " +
            "WHERE b.status = org.example.kalkulationsprogramm.domain.BelegStatus.VALIDIERT " +
            "AND ks.istFixkosten = true " +
            "AND (" +
            "  ((a.streckungJahre IS NULL OR a.streckungJahre <= 1) " +
            "   AND (a.streckungStartJahr IS NULL OR a.streckungStartJahr = :jahr))" +
            "  OR " +
            "  (a.streckungJahre > 1 AND a.streckungStartJahr IS NOT NULL " +
            "   AND :jahr >= a.streckungStartJahr " +
            "   AND :jahr < a.streckungStartJahr + a.streckungJahre)" +
            ")",
    )
    fun findAktiveFixkostenAnteileImJahr(@Param("jahr") jahr: Int): List<BelegKostenstellenAnteil>
}
